File: nodejs/tests/tof/vl53l0x_test_unit.js
'use strict';
const { I2CConnectionMock } = require('../../packages/periph/src/connection/i2c_mock');
const { VL53L0XMinimal, VL53L0XFull } = require('../../packages/periph/src/chips/tof/vl53l0x');

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else { console.log('FAIL', label); failed++; }
}

async function checkRejects(label, fn, type) {
    try {
        await fn();
        console.log(`FAIL ${label}: did not reject`); failed++;
    } catch (e) {
        checkTrue(label, !type || e instanceof type);
    }
}

const eq = (a, b) => Buffer.compare(Buffer.from(a), Buffer.from(b)) === 0;

// Page-aware VL53L0X simulator on top of the byte-slot mock. Registers
// written while 0xFF != 0 go to a separate per-page store, so the
// private-bank tuning writes don't clobber page-0 registers. Starting a
// ranging (or calibration) raises RESULT_INTERRUPT_STATUS; the interrupt
// clear drops it unless continuous mode is active. The SPAD-info handshake
// (page 7, 0x83) completes immediately.
class VL53L0XSim extends I2CConnectionMock {
    constructor() {
        super();
        this.page = 0;
        this.pages = new Map();
        this.continuous = false;
        this.log = [];
        this.setRegister(0xC0, [0xEE, 0xAA, 0x10]);
        this.setRegister(0x89, [0x00]);
        this.setRegister(0x60, [0x00]);
        this.setRegister(0x84, [0x11]);
        this.setRegister(0xB0, [0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF]);
        this.setRegister(0xF8, [0x00, 0x10]);
        // Result block: status 11, 10.0 SPADs, 5.0 MCPS signal, 0.5 MCPS ambient, 250 mm.
        this.setRegister(0x14, [11 << 3, 0x00, 0x0A, 0x00, 0x00, 0x00, 0x02, 0x80, 0x00, 0x40, 0x00, 0xFA]);
        this.pages.set('1:145', 0x3C);   // page 1, 0x91 stop variable
        this.pages.set('7:146', 0x85);   // page 7, 0x92 SPAD info: aperture, 5 SPADs
    }

    pageReg(page, reg) { return this.pages.get(`${page}:${reg}`); }

    reg(r) { return this.registers.get(r) || 0; }

    reg16(r) { return (this.reg(r) << 8) | this.reg(r + 1); }

    page0Writes(r) { return this.log.filter(([p, w]) => p === 0 && w.length >= 2 && w[0] === r).map(([, w]) => w); }

    logged(bytes) { return this.log.some(([, w]) => eq(w, bytes)); }

    async write(data) {
        const buf = Buffer.from(data);
        const reg = buf[0];
        if (reg === 0xFF) this.page = buf[1];
        this.log.push([this.page, buf]);
        if (this.page !== 0 && reg !== 0xFF) {
            for (let i = 1; i < buf.length; i++) this.pages.set(`${this.page}:${reg + i - 1}`, buf[i]);
            if (this.page === 7 && reg === 0x83 && buf[1] === 0x00) this.pages.set('7:131', 0x01);
            return;
        }
        await super.write(buf);
        if (reg === 0x00 && buf.length === 2) {
            const value = buf[1];
            if (value & 0x06) {
                this.continuous = true;
                this.registers.set(0x13, 0x04);
            } else if (value & 0x01) {
                if (this.continuous) this.continuous = false;
                else this.registers.set(0x13, 0x04);
            }
            this.registers.set(0x00, 0x00);
        } else if (reg === 0x0B && buf[1] === 0x01 && !this.continuous) {
            this.registers.set(0x13, 0x00);
        }
    }

    async writeRead(data, n) {
        if (this.page !== 0) {
            const out = Buffer.alloc(n);
            for (let i = 0; i < n; i++) out[i] = this.pageReg(this.page, data[0] + i) || 0;
            return out;
        }
        return super.writeRead(data, n);
    }
}

async function main() {
    // --- Identity check -----------------------------------------------------
    const bad = new VL53L0XSim();
    bad.registers.set(0xC0, 0xEF);
    await checkRejects('init rejects wrong model id', () => new VL53L0XMinimal(bad).init());

    // --- Initialization -----------------------------------------------------
    const mock = new VL53L0XSim();
    const sensor = new VL53L0XMinimal(mock);
    await sensor.init();
    checkTrue('init 2v8 mode', (mock.reg(0x89) & 0x01) === 0x01);
    checkTrue('init i2c standard mode', mock.registers.get(0x88) === 0x00);
    checkTrue('init stop variable', sensor._stopVariable === 0x3C);
    checkTrue('init signal checks disabled', eq(mock.page0Writes(0x60)[0], [0x60, 0x12]));
    checkTrue('init signal rate limit', eq(mock.page0Writes(0x44)[0], [0x44, 0x00, 0x20]));
    checkTrue('init spad map aperture 5',
        [0, 1, 2, 3, 4, 5].map((i) => mock.reg(0xB0 + i)).join() === [0x00, 0xF0, 0x01, 0, 0, 0].join());
    checkTrue('init ref en start select', mock.reg(0xB6) === 0xB4);
    checkTrue('init dynamic spad page1', mock.pageReg(1, 0x4E) === 0x2C && mock.pageReg(1, 0x4F) === 0x00);
    checkTrue('init tuning page0', mock.reg(0x46) === 0x25 && mock.reg(0x70) === 0x04);
    checkTrue('init tuning page1', mock.pageReg(1, 0x46) === 0x05);
    checkTrue('init gpio new sample', mock.reg(0x0A) === 0x04);
    checkTrue('init gpio active low', mock.reg(0x84) === 0x01);
    checkTrue('init sequence config', mock.reg(0x01) === 0xE8);
    const starts = mock.page0Writes(0x00).map((w) => w[1]);
    checkTrue('init vhv then phase calibration', starts.slice(-4).join() === [0x41, 0x00, 0x01, 0x00].join());
    checkTrue('init returns to page0', mock.page === 0);
    const seq = mock.page0Writes(0x01).map((w) => w[1]);
    checkTrue('init sequence order', seq.slice(-4).join() === [0xE8, 0x01, 0x02, 0xE8].join());

    // --- Single-shot ranging ------------------------------------------------
    mock.log = [];
    checkTrue('distance mm', (await sensor.distance()) === 250);
    checkTrue('range valid', await sensor.rangeValid());
    const preamble = [[0x80, 0x01], [0xFF, 0x01], [0x00, 0x00], [0x91, 0x3C], [0x00, 0x01], [0xFF, 0x00], [0x80, 0x00]];
    checkTrue('distance stop variable preamble', preamble.every((w, i) => eq(mock.log[i][1], w)));
    checkTrue('distance start', eq(mock.log[7][1], [0x00, 0x01]));
    checkTrue('distance clears interrupt', eq(mock.log[mock.log.length - 1][1], [0x0B, 0x01]));

    mock.setRegister(0x14, [4 << 3]);
    mock.setRegister(0x1E, [0x1F, 0xFF]);
    checkTrue('distance out of range raw', (await sensor.distance()) === 8191);
    checkTrue('range invalid', !(await sensor.rangeValid()));

    // --- Full: measurement record -------------------------------------------
    const fm = new VL53L0XSim();
    const full = new VL53L0XFull(fm);
    checkTrue('full is minimal', full instanceof VL53L0XMinimal);
    checkTrue('minimal has no full api', typeof VL53L0XMinimal.prototype.startContinuous === 'undefined');
    await full.distance();
    const m = await full.readMeasurement();
    checkTrue('measurement distance', m.distanceMm === 250);
    checkTrue('measurement status', m.rangeStatus === 11 && (await full.rangeStatus()) === 11);
    checkTrue('measurement signal rate', m.signalRateMcps === 5.0);
    checkTrue('measurement ambient rate', m.ambientRateMcps === 0.5);
    checkTrue('measurement spads', m.effectiveSpadCount === 10.0);

    // --- Continuous ranging -------------------------------------------------
    fm.log = [];
    await full.startContinuous();
    checkTrue('continuous back to back', eq(fm.log[fm.log.length - 1][1], [0x00, 0x02]));
    checkTrue('continuous stop variable', fm.logged([0x91, 0x3C]));
    checkTrue('data ready', await full.dataReady());
    checkTrue('read continuous', (await full.readContinuous()) === 250);
    await full.stopContinuous();
    const stop = [[0x00, 0x01], [0xFF, 0x01], [0x00, 0x00], [0x91, 0x00], [0x00, 0x01], [0xFF, 0x00]];
    checkTrue('stop continuous sequence', stop.every((w, i) => eq(fm.log[fm.log.length - 6 + i][1], w)));
    await full.startContinuous(100);
    checkTrue('timed period', [0, 1, 2, 3].map((i) => fm.reg(0x04 + i)).join() === [0x00, 0x00, 0x06, 0x40].join());
    checkTrue('timed start', eq(fm.log[fm.log.length - 1][1], [0x00, 0x04]));
    await full.stopContinuous();

    // --- Timing budget ------------------------------------------------------
    const budget = await full.timingBudget();
    checkTrue('default budget about 33ms', budget >= 32000 && budget <= 34000);
    await full.setTimingBudget(50000);
    checkTrue('budget roundtrip', Math.abs((await full.timingBudget()) - 50000) < 50);
    await checkRejects('budget rejects below min', () => full.setTimingBudget(19999), RangeError);

    // --- Signal rate --------------------------------------------------------
    await full.setSignalRateLimit(0.1);
    checkTrue('signal rate encode', fm.reg16(0x44) === 13);
    checkTrue('signal rate decode', (await full.signalRateLimit()) === 13 / 128);
    await checkRejects('signal rate rejects negative', () => full.setSignalRateLimit(-1), RangeError);

    // --- VCSEL periods ------------------------------------------------------
    checkTrue('vcsel pre default', (await full.vcselPulsePeriod('pre_range')) === 14);
    checkTrue('vcsel final default', (await full.vcselPulsePeriod('final_range')) === 10);
    await full.setVcselPulsePeriod('pre_range', 18);
    checkTrue('vcsel pre 18', (await full.vcselPulsePeriod('pre_range')) === 18 &&
        fm.reg(0x57) === 0x50 && fm.reg(0x56) === 0x08);
    await full.setVcselPulsePeriod('final_range', 14);
    checkTrue('vcsel final 14', (await full.vcselPulsePeriod('final_range')) === 14 && fm.reg(0x48) === 0x48 &&
        fm.reg(0x32) === 0x03 && fm.reg(0x30) === 0x07 && fm.pageReg(1, 0x30) === 0x20);
    checkTrue('vcsel keeps budget', Math.abs((await full.timingBudget()) - 50000) < 300);
    checkTrue('vcsel restores sequence', fm.reg(0x01) === 0xE8);
    await checkRejects('vcsel rejects odd', () => full.setVcselPulsePeriod('pre_range', 13), RangeError);
    await checkRejects('vcsel rejects type', () => full.setVcselPulsePeriod('mid', 10), RangeError);

    await full.setProfile('high_speed');
    checkTrue('profile high speed', (await full.vcselPulsePeriod('pre_range')) === 14 &&
        (await full.vcselPulsePeriod('final_range')) === 10 &&
        Math.abs((await full.timingBudget()) - 20000) < 50 && fm.reg16(0x44) === 32);
    await full.setProfile('long_range');
    checkTrue('profile long range', (await full.vcselPulsePeriod('pre_range')) === 18 &&
        (await full.vcselPulsePeriod('final_range')) === 14 && fm.reg16(0x44) === 13);
    await checkRejects('profile rejects unknown', () => full.setProfile('turbo'), RangeError);

    // --- Offset and crosstalk -----------------------------------------------
    await full.setOffset(-10.25);
    checkTrue('offset encode', fm.reg16(0x28) === ((-41) & 0x0FFF));
    checkTrue('offset decode', (await full.offset()) === -10.25);
    await full.setOffset(12.5);
    checkTrue('offset positive', (await full.offset()) === 12.5);
    await checkRejects('offset rejects range', () => full.setOffset(512), RangeError);
    await full.setCrosstalkCompensation(0.5);
    checkTrue('crosstalk encode', fm.reg16(0x20) === 4096);
    await full.setCrosstalkCompensation(0);
    checkTrue('crosstalk off', fm.reg16(0x20) === 0);
    await checkRejects('crosstalk rejects range', () => full.setCrosstalkCompensation(8), RangeError);

    // --- Recalibrate --------------------------------------------------------
    fm.log = [];
    await full.recalibrate();
    checkTrue('recalibrate vhv and phase', fm.logged([0x00, 0x41]) && fm.logged([0x01, 0x02]) && fm.reg(0x01) === 0xE8);

    // --- Thresholds, address, identification --------------------------------
    await full.setInterruptThresholds(100, 801);
    checkTrue('thresholds encode', fm.reg16(0x0E) === 50 && fm.reg16(0x0C) === 400);
    const th = await full.interruptThresholds();
    checkTrue('thresholds decode', th.lowMm === 100 && th.highMm === 800);
    await checkRejects('thresholds reject order', () => full.setInterruptThresholds(500, 100), RangeError);
    await full.setAddress(0x30);
    checkTrue('set address', fm.reg(0x8A) === 0x30);
    await checkRejects('address rejects range', () => full.setAddress(0x78), RangeError);
    checkTrue('model id', (await full.modelId()) === 0xEE);
    checkTrue('revision id', (await full.revisionId()) === 0x10);

    // --- Interrupt API ------------------------------------------------------
    await full.enableInterrupt(VL53L0XFull.SOURCE_OUT_OF_WINDOW);
    checkTrue('enable interrupt', fm.reg(0x0A) === 0x03);
    await full.disableInterrupt(VL53L0XFull.SOURCE_LEVEL_LOW);
    checkTrue('disable inactive source ignored', fm.reg(0x0A) === 0x03);
    await full.disableInterrupt(VL53L0XFull.SOURCE_OUT_OF_WINDOW);
    checkTrue('disable active source', fm.reg(0x0A) === 0x00);
    await full.enableInterrupt(VL53L0XFull.SOURCE_NEW_SAMPLE_READY);
    fm.registers.set(0x13, 0x03 | 0x08);
    checkTrue('poll interrupt value', (await full.pollInterrupt()) === VL53L0XFull.SOURCE_OUT_OF_WINDOW);
    checkTrue('poll interrupt clears', fm.reg(0x13) === 0x00);
    checkTrue('poll interrupt none', (await full.pollInterrupt()) === 0);
    await checkRejects('enable rejects range', () => full.enableInterrupt(5), RangeError);

    fm.registers.set(0x13, 0x04);
    const got = await new Promise((resolve) => {
        full.onInterrupt((status) => resolve(status));
    });
    await full.offInterrupt();
    checkTrue('on interrupt polling fallback', got === VL53L0XFull.SOURCE_NEW_SAMPLE_READY);

    console.log(`Passed: ${passed}, Failed: ${failed}`);
    console.log('===DONE===');
    process.exit(failed === 0 ? 0 : 1);
}

main().catch((e) => { console.error(e); process.exit(1); });

File: python/tests/tof/vl53l0x_test_unit.py
"""Unit test for the VL53L0X — runs without hardware using the I2C mock.

Verifies the model-ID check, the initialization sequence (2V8 I/O, stop
variable, reference-SPAD selection, tuning table, GPIO1 config, sequence
config, reference calibrations), single-shot ranging with the stop-variable
preamble, result-block decoding, continuous/timed ranging, the timing-budget
round trip, signal-rate limit, VCSEL periods, profiles, offset and crosstalk
encoding, thresholds, address change, and the interrupt API.
"""

from periph.connection.i2c_mock import I2CConnectionMock
from periph.chips.tof.vl53l0x import (
    VL53L0XMinimal, VL53L0XFull, PRE_RANGE, FINAL_RANGE,
    SOURCE_NEW_SAMPLE_READY, SOURCE_OUT_OF_WINDOW, SOURCE_LEVEL_LOW)

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


def raises(fn, exc=ValueError):
    try:
        fn()
    except exc:
        return True
    return False


class VL53L0XSim(I2CConnectionMock):
    """Page-aware VL53L0X simulator on top of the byte-slot mock.

    Registers written while 0xFF != 0 go to a separate per-page store, so the
    private-bank tuning writes don't clobber page-0 registers. Starting a
    ranging (or calibration) raises RESULT_INTERRUPT_STATUS; the interrupt
    clear drops it unless continuous mode is active. The SPAD-info handshake
    (page 7, 0x83) completes immediately.
    """

    def __init__(self):
        super().__init__()
        self.page = 0
        self.pages = {}
        self.continuous = False
        self.log = []
        self.set_register(0xC0, 0xEE, 0xAA, 0x10)
        self.set_register(0x89, 0x00)
        self.set_register(0x60, 0x00)
        self.set_register(0x84, 0x11)
        self.set_register(0xB0, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF)
        self.set_register(0xF8, 0x00, 0x10)
        # Result block: status 11, 10.0 SPADs, 5.0 MCPS signal, 0.5 MCPS ambient, 250 mm.
        self.set_register(0x14, 11 << 3, 0x00, 0x0A, 0x00, 0x00, 0x00,
                          0x02, 0x80, 0x00, 0x40, 0x00, 0xFA)
        self.pages[(1, 0x91)] = 0x3C
        self.pages[(7, 0x92)] = 0x85

    def write(self, data):
        data = bytes(data)
        reg = data[0]
        if reg == 0xFF:
            self.page = data[1]
        self.log.append((self.page, data))
        if self.page != 0 and reg != 0xFF:
            self.writes.append(data)
            for i, value in enumerate(data[1:]):
                self.pages[(self.page, reg + i)] = value
            if self.page == 7 and reg == 0x83 and data[1] == 0x00:
                self.pages[(7, 0x83)] = 0x01
            return
        super().write(data)
        if reg == 0x00 and len(data) == 2:
            value = data[1]
            if value & 0x06:
                self.continuous = True
                self.registers[0x13] = 0x04
            elif value & 0x01:
                if self.continuous:
                    self.continuous = False
                else:
                    self.registers[0x13] = 0x04
            self.registers[0x00] = 0x00
        elif reg == 0x0B and data[1] == 0x01 and not self.continuous:
            self.registers[0x13] = 0x00

    def write_read(self, data, n):
        if self.page != 0:
            self.writes.append(bytes(data))
            return bytes(self.pages.get((self.page, data[0] + i), 0) for i in range(n))
        return super().write_read(data, n)


def page0_writes(mock, reg):
    return [w for p, w in mock.log if p == 0 and len(w) >= 2 and w[0] == reg]


def reg16(mock, reg):
    return (mock.registers.get(reg, 0) << 8) | mock.registers.get(reg + 1, 0)


# --- Identity check -------------------------------------------------------
bad = VL53L0XSim()
bad.registers[0xC0] = 0xEF
check_true('init_rejects_wrong_model_id', raises(lambda: VL53L0XMinimal(bad)))

# --- Initialization -------------------------------------------------------
mock = VL53L0XSim()
sensor = VL53L0XMinimal(mock)
check_true('init_2v8_mode', mock.registers[0x89] & 0x01 == 0x01)
check_true('init_i2c_standard_mode', mock.registers[0x88] == 0x00)
check_true('init_stop_variable', sensor._stop_variable == 0x3C)
check_true('init_signal_checks_disabled', page0_writes(mock, 0x60)[0] == bytes([0x60, 0x12]))
check_true('init_signal_rate_limit', reg16(mock, 0x44) == 0x0020)
check_true('init_spad_map_aperture_5', [mock.registers[0xB0 + i] for i in range(6)]
           == [0x00, 0xF0, 0x01, 0x00, 0x00, 0x00])
check_true('init_ref_en_start_select', mock.registers[0xB6] == 0xB4)
check_true('init_dynamic_spad_page1', mock.pages[(1, 0x4E)] == 0x2C and mock.pages[(1, 0x4F)] == 0x00)
check_true('init_tuning_page0', mock.registers[0x46] == 0x25 and mock.registers[0x70] == 0x04)
check_true('init_tuning_page1', mock.pages[(1, 0x46)] == 0x05)
check_true('init_gpio_new_sample', mock.registers[0x0A] == 0x04)
check_true('init_gpio_active_low', mock.registers[0x84] == 0x01)
check_true('init_sequence_config', mock.registers[0x01] == 0xE8)
starts = [w[1] for w in page0_writes(mock, 0x00)]
check_true('init_vhv_then_phase_calibration', starts[-4:] == [0x41, 0x00, 0x01, 0x00])
check_true('init_returns_to_page0', mock.page == 0)
seq_writes = [w[1] for w in page0_writes(mock, 0x01)]
check_true('init_sequence_order', seq_writes[-4:] == [0xE8, 0x01, 0x02, 0xE8])

# --- Single-shot ranging --------------------------------------------------
mock.writes.clear()
d = sensor.distance()
check_true('distance_mm', d == 250)
check_true('range_valid', sensor.range_valid())
pre = mock.writes[:7]
check_true('distance_stop_variable_preamble', pre == [
    bytes([0x80, 0x01]), bytes([0xFF, 0x01]), bytes([0x00, 0x00]), bytes([0x91, 0x3C]),
    bytes([0x00, 0x01]), bytes([0xFF, 0x00]), bytes([0x80, 0x00])])
check_true('distance_start', mock.writes[7] == bytes([0x00, 0x01]))
check_true('distance_clears_interrupt', mock.writes[-1] == bytes([0x0B, 0x01]))

mock.set_register(0x14, 4 << 3)
mock.set_register(0x1E, 0x1F, 0xFF)
check_true('distance_out_of_range_raw', sensor.distance() == 8191)
check_true('range_invalid', not sensor.range_valid())

# --- Full: measurement record ---------------------------------------------
mock = VL53L0XSim()
full = VL53L0XFull(mock)
check_true('full_is_minimal', isinstance(full, VL53L0XMinimal))
check_true('minimal_has_no_full_api', not hasattr(VL53L0XMinimal, 'start_continuous'))
full.distance()
m = full.read_measurement()
check_true('measurement_distance', m['distance_mm'] == 250)
check_true('measurement_status', m['range_status'] == 11 and full.range_status() == 11)
check_true('measurement_signal_rate', m['signal_rate_mcps'] == 5.0)
check_true('measurement_ambient_rate', m['ambient_rate_mcps'] == 0.5)
check_true('measurement_spads', m['effective_spad_count'] == 10.0)

# --- Continuous ranging ---------------------------------------------------
mock.writes.clear()
full.start_continuous()
check_true('continuous_back_to_back', mock.writes[-1] == bytes([0x00, 0x02]))
check_true('continuous_stop_variable', bytes([0x91, 0x3C]) in mock.writes)
check_true('data_ready', full.data_ready())
check_true('read_continuous', full.read_continuous() == 250)
full.stop_continuous()
check_true('stop_continuous_sequence', mock.writes[-6:] == [
    bytes([0x00, 0x01]), bytes([0xFF, 0x01]), bytes([0x00, 0x00]), bytes([0x91, 0x00]),
    bytes([0x00, 0x01]), bytes([0xFF, 0x00])])
full.start_continuous(100)
check_true('timed_period', [mock.registers[0x04 + i] for i in range(4)] == [0x00, 0x00, 0x06, 0x40])
check_true('timed_start', mock.writes[-1] == bytes([0x00, 0x04]))
full.stop_continuous()

# --- Timing budget --------------------------------------------------------
budget = full.timing_budget()
check_true('default_budget_about_33ms', 32000 <= budget <= 34000)
full.set_timing_budget(50000)
check_true('budget_roundtrip', abs(full.timing_budget() - 50000) < 50)
check_true('budget_rejects_below_min', raises(lambda: full.set_timing_budget(19999)))

# --- Signal rate ----------------------------------------------------------
full.set_signal_rate_limit(0.1)
check_true('signal_rate_encode', reg16(mock, 0x44) == 13)
check_true('signal_rate_decode', full.signal_rate_limit() == 13 / 128)
check_true('signal_rate_rejects_negative', raises(lambda: full.set_signal_rate_limit(-1)))

# --- VCSEL periods --------------------------------------------------------
check_true('vcsel_pre_default', full.vcsel_pulse_period(PRE_RANGE) == 14)
check_true('vcsel_final_default', full.vcsel_pulse_period(FINAL_RANGE) == 10)
full.set_vcsel_pulse_period(PRE_RANGE, 18)
check_true('vcsel_pre_18', full.vcsel_pulse_period(PRE_RANGE) == 18
           and mock.registers[0x57] == 0x50 and mock.registers[0x56] == 0x08)
full.set_vcsel_pulse_period(FINAL_RANGE, 14)
check_true('vcsel_final_14', full.vcsel_pulse_period(FINAL_RANGE) == 14
           and mock.registers[0x48] == 0x48 and mock.registers[0x32] == 0x03
           and mock.registers[0x30] == 0x07 and mock.pages[(1, 0x30)] == 0x20)
check_true('vcsel_keeps_budget', abs(full.timing_budget() - 50000) < 300)
check_true('vcsel_restores_sequence', mock.registers[0x01] == 0xE8)
check_true('vcsel_rejects_odd', raises(lambda: full.set_vcsel_pulse_period(PRE_RANGE, 13)))
check_true('vcsel_rejects_type', raises(lambda: full.set_vcsel_pulse_period('mid', 10)))

full.set_profile('high_speed')
check_true('profile_high_speed', full.vcsel_pulse_period(PRE_RANGE) == 14
           and full.vcsel_pulse_period(FINAL_RANGE) == 10
           and abs(full.timing_budget() - 20000) < 50 and reg16(mock, 0x44) == 32)
full.set_profile('long_range')
check_true('profile_long_range', full.vcsel_pulse_period(PRE_RANGE) == 18
           and full.vcsel_pulse_period(FINAL_RANGE) == 14 and reg16(mock, 0x44) == 13)
check_true('profile_rejects_unknown', raises(lambda: full.set_profile('turbo')))

# --- Offset and crosstalk -------------------------------------------------
full.set_offset(-10.25)
check_true('offset_encode', reg16(mock, 0x28) == ((-41) & 0x0FFF))
check_true('offset_decode', full.offset() == -10.25)
full.set_offset(12.5)
check_true('offset_positive', full.offset() == 12.5)
check_true('offset_rejects_range', raises(lambda: full.set_offset(512.0)))
full.set_crosstalk_compensation(0.5)
check_true('crosstalk_encode', reg16(mock, 0x20) == 4096)
full.set_crosstalk_compensation(0)
check_true('crosstalk_off', reg16(mock, 0x20) == 0)
check_true('crosstalk_rejects_range', raises(lambda: full.set_crosstalk_compensation(8.0)))

# --- Recalibrate ----------------------------------------------------------
mock.writes.clear()
full.recalibrate()
check_true('recalibrate_vhv_and_phase', bytes([0x00, 0x41]) in mock.writes
           and bytes([0x01, 0x02]) in mock.writes and mock.registers[0x01] == 0xE8)

# --- Thresholds, address, identification ----------------------------------
full.set_interrupt_thresholds(100, 801)
check_true('thresholds_encode', reg16(mock, 0x0E) == 50 and reg16(mock, 0x0C) == 400)
check_true('thresholds_decode', full.interrupt_thresholds() == (100, 800))
check_true('thresholds_reject_order', raises(lambda: full.set_interrupt_thresholds(500, 100)))
full.set_address(0x30)
check_true('set_address', mock.registers[0x8A] == 0x30)
check_true('address_rejects_range', raises(lambda: full.set_address(0x78)))
check_true('model_id', full.model_id() == 0xEE)
check_true('revision_id', full.revision_id() == 0x10)

# --- Interrupt API --------------------------------------------------------
full.enable_interrupt(SOURCE_OUT_OF_WINDOW)
check_true('enable_interrupt', mock.registers[0x0A] == 0x03)
full.disable_interrupt(SOURCE_LEVEL_LOW)
check_true('disable_inactive_source_ignored', mock.registers[0x0A] == 0x03)
full.disable_interrupt(SOURCE_OUT_OF_WINDOW)
check_true('disable_active_source', mock.registers[0x0A] == 0x00)
full.enable_interrupt(SOURCE_NEW_SAMPLE_READY)
mock.registers[0x13] = 0x03 | 0x08
check_true('poll_interrupt_value', full.poll_interrupt() == SOURCE_OUT_OF_WINDOW)
check_true('poll_interrupt_clears', mock.registers[0x13] == 0x00)
check_true('poll_interrupt_none', full.poll_interrupt() == 0)
check_true('enable_rejects_range', raises(lambda: full.enable_interrupt(5)))

print('Passed: {}, Failed: {}'.format(passed, failed))
print('===DONE===')
